File: scala/src/test/scala/jbaldassari/leetcode/topinterview150/arraystring/removeduplicatessortedarray/RemoveDuplicatesFromSortedArrayTest.scala
package jbaldassari.leetcode.topinterview150.arraystring.removeduplicatessortedarray

class RemoveDuplicatesFromSortedArrayTest extends munit.FunSuite {
  test("example 1") {
    evaluate(Array(1, 1, 2), 2, Array(1, 2))
  }
  test("example 2") {
    evaluate(Array(0, 0, 1, 1, 1, 2, 2, 3, 3, 4), 5, Array(0, 1, 2, 3, 4))
  }
  test("array of size 1: no removals") {
    evaluate(Array(1),  1, Array(1))
  }
  test("array of size 1: no removals") {
    evaluate(Array(1), 1, Array(1))
  }
  test("array of size 2: all duplicates") {
    evaluate(Array(1, 1), 1, Array(1))
  }
  test("array of size 2: no duplicates") {
    evaluate(Array(1, 2), 2, Array(1, 2))
  }
  test("array of size 3: all duplicates") {
    evaluate(Array(1, 1, 1), 1, Array(1))
  }
  test("array of size 3: dupes at beginning") {
    evaluate(Array(0, 0, 1), 2, Array(0, 1))
  }
  test("array of size 3: dupes at end") {
    evaluate(Array(0, 1, 1), 2, Array(0, 1))
  }
  test("array of size 4: dupes in middle") {
    evaluate(Array(0, 1, 1, 2), 3, Array(0, 1, 2))
  }
  def evaluate(nums: Array[Int], expectedOutput: Int, expectedNums: Array[Int]): Unit = {
    val numsCopy = nums.clone()
    val result = Solution.removeDuplicates(numsCopy)
    assertEquals(result, expectedOutput)
    assertEquals(numsCopy.toList.take(expectedNums.size), expectedNums.toList)
  }
}

File: typescript/src/TopInterview150/Array-String/RemoveDuplicatesFromSortedArray/RemoveDuplicatesFromSortedArray.test.ts
import {removeDuplicates} from './RemoveDuplicatesFromSortedArray';

describe('Remove Duplicates from Sorted Array', () => {
  it('example 1', () => {
    evaluate([1, 1, 2], 2, [1, 2]);
  });
  it('example 2', () => {
    evaluate([0, 0, 1, 1, 1, 2, 2, 3, 3, 4], 5, [0, 1, 2, 3, 4]);
  });
  it('array of size 1: no removals', () => {
    evaluate([1], 1, [1]);
  });
  it('array of size 2: all duplicates', () => {
    evaluate([1, 1], 1, [1]);
  });
  it('array of size 2: no duplicates', () => {
    evaluate([1, 2], 2, [1, 2]);
  });
  it('array of size 3: all duplicates', () => {
    evaluate([1, 1, 1], 1, [1]);
  });
  it('array of size 3: dupes at beginning', () => {
    evaluate([0, 0, 1], 2, [0, 1]);
  });
  it('array of size 3: dupes at end', () => {
    evaluate([0, 1, 1], 2, [0, 1]);
  });
  it('array of size 4: dupes in middle', () => {
    evaluate([0, 1, 1, 2], 3, [0, 1, 2]);
  });
});

function evaluate(nums: number[], expectedOutput: number, expectedNums: number[]): void {
  const numsCopy = [...nums];
  const output = removeDuplicates(numsCopy);
  expect(output).toBe(expectedOutput);
  expect(numsCopy.slice(0, output)).toStrictEqual(expectedNums);
}
